File: ts/src/policy.ts
/**
 * AgentGuard policy engine: deny-by-default, least-privilege tool authorization.
 *
 * Every agent tool call is evaluated against a policy before execution. Decisions
 * are ALLOW, DENY, or APPROVE (human-in-the-loop for high-risk actions).
 */

/** The outcome of evaluating a tool call against a policy. */
export enum Decision {
  Allow = "allow",
  Deny = "deny",
  Approve = "approve",
}

/** Arbitrary tool-call arguments, mirroring the reference's dynamic dict. */
export type Args = Record<string, unknown>;

/** A single request by the agent to invoke a named tool with arguments. */
export class ToolCall {
  constructor(
    public readonly tool: string,
    public readonly args: Args = {},
  ) {}
}

/** The constraints attached to one allowed tool. Defaults deny everything. */
export interface ToolPolicy {
  /** The tool this policy governs. */
  tool: string;
  /** Whether the tool is allowed at all. */
  allow: boolean;
  /** Glob patterns a `path` argument must match. */
  pathAllow: string[];
  /** Glob patterns a `path` argument must not match. */
  pathDeny: string[];
  /** Exact domains a `domain` argument must be one of. */
  domainAllow: string[];
  /** Whether the tool requires human approval. */
  requireApproval: boolean;
}

/** Build a {@link ToolPolicy}, defaulting every unset constraint to deny/empty. */
export function toolPolicy(
  tool: string,
  opts: Partial<Omit<ToolPolicy, "tool">> = {},
): ToolPolicy {
  return {
    tool,
    allow: opts.allow ?? false,
    pathAllow: opts.pathAllow ?? [],
    pathDeny: opts.pathDeny ?? [],
    domainAllow: opts.domainAllow ?? [],
    requireApproval: opts.requireApproval ?? false,
  };
}

/** The snake_case spec for a single tool, as loaded by {@link Policy.fromDict}. */
export interface ToolSpec {
  tool: string;
  allow?: boolean;
  path_allow?: string[];
  path_deny?: string[];
  domain_allow?: string[];
  require_approval?: boolean;
}

/** A whole-policy spec, e.g. parsed from a JSON policy file. */
export interface PolicySpec {
  tools?: ToolSpec[];
}

/** Deny-by-default: only explicitly allowed tools/constraints pass. */
export class Policy {
  readonly tools: Map<string, ToolPolicy>;

  constructor(tools: ToolPolicy[] = []) {
    this.tools = new Map(tools.map((t) => [t.tool, t]));
  }

  /** Build a policy from a plain spec object (mirrors the Python `from_dict`). */
  static fromDict(spec: PolicySpec): Policy {
    const tools = (spec.tools ?? []).map((t) =>
      toolPolicy(t.tool, {
        allow: t.allow ?? false,
        pathAllow: t.path_allow ?? [],
        pathDeny: t.path_deny ?? [],
        domainAllow: t.domain_allow ?? [],
        requireApproval: t.require_approval ?? false,
      }),
    );
    return new Policy(tools);
  }

  /** Evaluate a tool call, returning the decision and a human-readable reason. */
  evaluate(call: ToolCall): [Decision, string] {
    const tp = this.tools.get(call.tool);
    if (!tp || !tp.allow) {
      return [
        Decision.Deny,
        `tool '${call.tool}' not in allow-list (deny-by-default)`,
      ];
    }

    const path = argString(call.args.path);
    // Deny-by-default extends to constrained arguments: a restricted tool called
    // with no path can't be proven in-bounds, so we refuse rather than allow.
    if (tp.pathAllow.length > 0 && path === "") {
      return [
        Decision.Deny,
        `tool '${call.tool}' requires a path within its allowed set`,
      ];
    }
    if (path !== "") {
      for (const pattern of tp.pathDeny) {
        if (fnmatch(path, pattern)) {
          return [
            Decision.Deny,
            `path '${path}' matches deny pattern '${pattern}'`,
          ];
        }
      }
      if (tp.pathAllow.length > 0 && !tp.pathAllow.some((p) => fnmatch(path, p))) {
        return [Decision.Deny, `path '${path}' not in allowed paths`];
      }
    }

    const domain = argString(call.args.domain);
    if (tp.domainAllow.length > 0 && domain === "") {
      return [
        Decision.Deny,
        `tool '${call.tool}' requires a domain within its allowed set`,
      ];
    }
    if (domain !== "" && tp.domainAllow.length > 0 && !tp.domainAllow.includes(domain)) {
      return [Decision.Deny, `domain '${domain}' not in allow-list`];
    }

    if (tp.requireApproval) {
      return [Decision.Approve, `tool '${call.tool}' requires human approval`];
    }
    return [Decision.Allow, "ok"];
  }
}

/** Mirror Python's `str(x) if x is not None else ""`: null/undefined become "". */
function argString(value: unknown): string {
  if (value === undefined || value === null) {
    return "";
  }
  return String(value);
}

// --- glob matching with Python fnmatch semantics (case-sensitive, full match) ---

type Tok =
  | { kind: "star" }
  | { kind: "any" }
  | { kind: "lit"; ch: string }
  | { kind: "class"; neg: boolean; ranges: Array<[number, number]> };

/**
 * Match `name` against a shell glob using Python `fnmatch` semantics: case
 * sensitive, whole-string, `*` matches any run (including `/`), `?` any single
 * char, and `[...]`/`[!...]` character classes.
 */
export function fnmatch(name: string, pattern: string): boolean {
  const toks = compile(pattern);
  const chars = Array.from(name);

  let i = 0;
  let j = 0;
  let star = -1;
  let starI = 0;
  while (i < chars.length) {
    if (j < toks.length && toks[j].kind !== "star" && tokMatches(toks[j], chars[i])) {
      i += 1;
      j += 1;
    } else if (j < toks.length && toks[j].kind === "star") {
      star = j;
      starI = i;
      j += 1;
    } else if (star !== -1) {
      j = star + 1;
      starI += 1;
      i = starI;
    } else {
      return false;
    }
  }
  while (j < toks.length && toks[j].kind === "star") {
    j += 1;
  }
  return j === toks.length;
}

function tokMatches(tok: Tok, c: string): boolean {
  switch (tok.kind) {
    case "any":
      return true;
    case "lit":
      return tok.ch === c;
    case "class": {
      const cc = c.codePointAt(0) ?? -1;
      const inside = tok.ranges.some(([lo, hi]) => lo <= cc && cc <= hi);
      return inside !== tok.neg;
    }
    case "star":
      return false;
  }
}

function compile(pattern: string): Tok[] {
  const p = Array.from(pattern);
  const toks: Tok[] = [];
  let i = 0;
  while (i < p.length) {
    const c = p[i];
    if (c === "*") {
      toks.push({ kind: "star" });
      i += 1;
    } else if (c === "?") {
      toks.push({ kind: "any" });
      i += 1;
    } else if (c === "[") {
      let j = i + 1;
      if (j < p.length && p[j] === "!") {
        j += 1;
      }
      if (j < p.length && p[j] === "]") {
        j += 1;
      }
      while (j < p.length && p[j] !== "]") {
        j += 1;
      }
      if (j >= p.length) {
        // No closing bracket: treat '[' as a literal.
        toks.push({ kind: "lit", ch: "[" });
        i += 1;
      } else {
        toks.push(parseClass(p.slice(i + 1, j)));
        i = j + 1;
      }
    } else {
      toks.push({ kind: "lit", ch: c });
      i += 1;
    }
  }
  return toks;
}

function parseClass(body: string[]): Tok {
  let neg = false;
  let idx = 0;
  if (body.length > 0 && body[0] === "!") {
    neg = true;
    idx = 1;
  }
  const ranges: Array<[number, number]> = [];
  while (idx < body.length) {
    if (idx + 2 < body.length && body[idx + 1] === "-") {
      ranges.push([body[idx].codePointAt(0) ?? -1, body[idx + 2].codePointAt(0) ?? -1]);
      idx += 3;
    } else {
      const cp = body[idx].codePointAt(0) ?? -1;
      ranges.push([cp, cp]);
      idx += 1;
    }
  }
  return { kind: "class", neg, ranges };
}

File: rust/src/policy.rs
//! AgentGuard policy engine: deny-by-default, least-privilege tool authorization.
//!
//! Every agent tool call is evaluated against a policy before execution. Decisions
//! are ALLOW, DENY, or APPROVE (human-in-the-loop for high-risk actions).

use std::collections::HashMap;

/// The outcome of evaluating a tool call against a policy.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Decision {
    /// Permit the tool call to execute.
    Allow,
    /// Block the tool call.
    Deny,
    /// Require human approval before the call may execute.
    Approve,
}

impl Decision {
    /// The lowercase wire string stored in the audit log.
    pub fn value(self) -> &'static str {
        match self {
            Decision::Allow => "allow",
            Decision::Deny => "deny",
            Decision::Approve => "approve",
        }
    }
}

/// A single argument value in a tool call. Mirrors the reference's dynamically
/// typed args: a string, an integer, an explicit null, or some other object.
#[derive(Debug, Clone)]
pub enum ArgValue {
    /// A string argument.
    Str(String),
    /// An integer argument.
    Int(i64),
    /// An explicit null (treated the same as a missing argument).
    Null,
    /// Any other object; stringifies to a non-empty placeholder.
    Other,
}

impl ArgValue {
    fn as_arg_string(&self) -> String {
        match self {
            ArgValue::Str(s) => s.clone(),
            ArgValue::Int(n) => n.to_string(),
            ArgValue::Null => String::new(),
            ArgValue::Other => "<object>".to_string(),
        }
    }
}

impl From<&str> for ArgValue {
    fn from(s: &str) -> Self {
        ArgValue::Str(s.to_string())
    }
}

impl From<String> for ArgValue {
    fn from(s: String) -> Self {
        ArgValue::Str(s)
    }
}

impl From<i64> for ArgValue {
    fn from(n: i64) -> Self {
        ArgValue::Int(n)
    }
}

/// A single request by the agent to invoke a named tool with arguments.
#[derive(Debug, Clone)]
pub struct ToolCall {
    /// The tool name.
    pub tool: String,
    /// The call arguments.
    pub args: HashMap<String, ArgValue>,
}

impl ToolCall {
    /// Build a tool call with no arguments.
    pub fn new(tool: impl Into<String>) -> Self {
        ToolCall {
            tool: tool.into(),
            args: HashMap::new(),
        }
    }

    /// Build a tool call with the given arguments.
    pub fn with_args(tool: impl Into<String>, args: HashMap<String, ArgValue>) -> Self {
        ToolCall {
            tool: tool.into(),
            args,
        }
    }

    /// Mirror the reference's `str(args.get(key))` with `None -> ""`: a missing or
    /// null value becomes the empty string; any other value is stringified.
    fn arg_string(&self, key: &str) -> String {
        self.args
            .get(key)
            .map(ArgValue::as_arg_string)
            .unwrap_or_default()
    }
}

/// The constraints attached to one allowed tool. The default value denies
/// everything: `allow` is false and every constraint list is empty.
#[derive(Debug, Clone, Default)]
pub struct ToolPolicy {
    /// The tool this policy governs.
    pub tool: String,
    /// Whether the tool is allowed at all.
    pub allow: bool,
    /// Glob patterns a `path` argument must match.
    pub path_allow: Vec<String>,
    /// Glob patterns a `path` argument must not match.
    pub path_deny: Vec<String>,
    /// Exact domains a `domain` argument must be one of.
    pub domain_allow: Vec<String>,
    /// Whether the tool requires human approval.
    pub require_approval: bool,
}

/// Deny-by-default: only explicitly allowed tools, within their declared
/// constraints, ever pass.
#[derive(Debug, Clone, Default)]
pub struct Policy {
    /// The allow-table, keyed by tool name.
    pub tools: HashMap<String, ToolPolicy>,
}

impl Policy {
    /// Build a policy from a list of tool policies, keyed by tool name.
    pub fn new(tools: Vec<ToolPolicy>) -> Self {
        let mut map = HashMap::new();
        for t in tools {
            map.insert(t.tool.clone(), t);
        }
        Policy { tools: map }
    }

    /// Evaluate a tool call and return the decision plus a human-readable reason.
    pub fn evaluate(&self, call: &ToolCall) -> (Decision, String) {
        let tp = match self.tools.get(&call.tool) {
            Some(tp) if tp.allow => tp,
            _ => {
                return (
                    Decision::Deny,
                    format!("tool '{}' not in allow-list (deny-by-default)", call.tool),
                )
            }
        };

        let path = call.arg_string("path");
        // Deny-by-default extends to constrained arguments: if a tool is restricted
        // to certain paths but the call provides none, we can't prove it's in
        // bounds, so we refuse rather than fall through to allow.
        if !tp.path_allow.is_empty() && path.is_empty() {
            return (
                Decision::Deny,
                format!(
                    "tool '{}' requires a path within its allowed set",
                    call.tool
                ),
            );
        }
        if !path.is_empty() {
            for pattern in &tp.path_deny {
                if fnmatch(&path, pattern) {
                    return (
                        Decision::Deny,
                        format!("path '{path}' matches deny pattern '{pattern}'"),
                    );
                }
            }
            if !tp.path_allow.is_empty() && !tp.path_allow.iter().any(|p| fnmatch(&path, p)) {
                return (
                    Decision::Deny,
                    format!("path '{path}' not in allowed paths"),
                );
            }
        }

        let domain = call.arg_string("domain");
        if !tp.domain_allow.is_empty() && domain.is_empty() {
            return (
                Decision::Deny,
                format!(
                    "tool '{}' requires a domain within its allowed set",
                    call.tool
                ),
            );
        }
        if !domain.is_empty() && !tp.domain_allow.is_empty() && !tp.domain_allow.contains(&domain) {
            return (
                Decision::Deny,
                format!("domain '{domain}' not in allow-list"),
            );
        }

        if tp.require_approval {
            return (
                Decision::Approve,
                format!("tool '{}' requires human approval", call.tool),
            );
        }
        (Decision::Allow, "ok".to_string())
    }
}

// --- glob matching with Python fnmatch semantics (case-sensitive, full match) ---

enum Tok {
    Star,
    Any,
    Lit(char),
    Class {
        neg: bool,
        ranges: Vec<(char, char)>,
    },
}

/// Match `name` against a shell glob using Python `fnmatch` semantics: case
/// sensitive, whole-string, `*` matches any run (including `/`), `?` any single
/// char, and `[...]`/`[!...]` character classes.
fn fnmatch(name: &str, pattern: &str) -> bool {
    let toks = compile(&pattern.chars().collect::<Vec<char>>());
    let name: Vec<char> = name.chars().collect();

    let (mut i, mut j) = (0usize, 0usize);
    let mut star: Option<usize> = None;
    let mut star_i = 0usize;
    while i < name.len() {
        if j < toks.len() && !matches!(toks[j], Tok::Star) && tok_matches(&toks[j], name[i]) {
            i += 1;
            j += 1;
        } else if j < toks.len() && matches!(toks[j], Tok::Star) {
            star = Some(j);
            star_i = i;
            j += 1;
        } else if let Some(sj) = star {
            j = sj + 1;
            star_i += 1;
            i = star_i;
        } else {
            return false;
        }
    }
    while j < toks.len() && matches!(toks[j], Tok::Star) {
        j += 1;
    }
    j == toks.len()
}

fn tok_matches(tok: &Tok, c: char) -> bool {
    match tok {
        Tok::Any => true,
        Tok::Lit(l) => *l == c,
        Tok::Class { neg, ranges } => {
            let inside = ranges.iter().any(|(lo, hi)| *lo <= c && c <= *hi);
            inside != *neg
        }
        Tok::Star => false,
    }
}

fn compile(pattern: &[char]) -> Vec<Tok> {
    let mut toks = Vec::new();
    let mut i = 0;
    while i < pattern.len() {
        match pattern[i] {
            '*' => {
                toks.push(Tok::Star);
                i += 1;
            }
            '?' => {
                toks.push(Tok::Any);
                i += 1;
            }
            '[' => {
                let mut j = i + 1;
                if j < pattern.len() && pattern[j] == '!' {
                    j += 1;
                }
                if j < pattern.len() && pattern[j] == ']' {
                    j += 1;
                }
                while j < pattern.len() && pattern[j] != ']' {
                    j += 1;
                }
                if j >= pattern.len() {
                    // No closing bracket: treat '[' as a literal.
                    toks.push(Tok::Lit('['));
                    i += 1;
                } else {
                    toks.push(parse_class(&pattern[i + 1..j]));
                    i = j + 1;
                }
            }
            c => {
                toks.push(Tok::Lit(c));
                i += 1;
            }
        }
    }
    toks
}

fn parse_class(body: &[char]) -> Tok {
    let mut neg = false;
    let mut idx = 0;
    if !body.is_empty() && body[0] == '!' {
        neg = true;
        idx = 1;
    }
    let mut ranges = Vec::new();
    while idx < body.len() {
        if idx + 2 < body.len() && body[idx + 1] == '-' {
            ranges.push((body[idx], body[idx + 2]));
            idx += 3;
        } else {
            ranges.push((body[idx], body[idx]));
            idx += 1;
        }
    }
    Tok::Class { neg, ranges }
}
